src/hotspot/share/gc/shared/barrierSetNMethod.cpp

```diff
@@ -62,11 +65,17 @@ bool BarrierSetNMethod::supports_entry_barrier(nmethod* nm) {
     return false;
   }
 
-  if (!nm->is_native_method() && !nm->is_compiled_by_c2() && !nm->is_compiled_by_c1()) {
-    return false;
+  if (nm->is_native_method() || nm->is_compiled_by_c2() || nm->is_compiled_by_c1()) {
+    return true;
   }
```

Comment:
I wonder why `is_native_method` automatically implies `supports_entry_barrier == true`.
Apparently JVMCI is not supposed to compile native methods (without entry barrier), however, this seems too restrictive to me.

Should not we swap the order of the checks? E.g.
```
  if (nm->is_compiled_by_jvmci()) {
    return nm->jvmci_nmethod_data()->has_entry_barrier();
  }

  if (nm->is_native_method() || nm->is_compiled_by_c2() || nm->is_compiled_by_c1()) {
    return true;
  }
```